### PR TITLE
[Beta][Bug] Major/minor ME bug fixes

### PR DESCRIPTION
**Pull request by ben-lear**
Tracking issue from here: https://github.com/pagefaultgames/pokerogue/issues/4372

## Fixed Issues

- **[FIXED]** Mythicals/Legendaries should no longer improperly spawn in Safari Zone and other encounters
- **[FIXED]** Fixed RNG generation with egg moves in Weird Dream and a few other places that used `randSeedShuffle`
- **[FIXED]** Weird Dream: The game properly adds the favored move to a real index for a Pokemon with less than 4 moves
- **[FIXED]** Clowning around type randomization has been fixed and is not able to be manipulated
- **[FIXED]** Clowning Around will safely exit in a (theoretically impossible) scenario where the game runs out of items to generate
- **[FIXED]** Absolute Avarice: Text for Reviver Seed rewards shows properly if the entire team already has Reviver Seeds
- **[FIXED]** Safari Zone text is fixed with new styling
- **[FIXED]** Can no longer flee from certain wild ME encounters that are intended to be forced
- **[FIXED]** "Rental" Pokemon from Weird Dream are no longer able to be falsely unlocked in dex later via other methods (Training Session, form changes, etc.)
- **[FIXED]** Macho Brace cannot affect a Wonder Guard pokemon's HP
- **[FIXED]** Golden Bug Net no longer works in END biome nor replaces boss encounters
- **[FIXED]** Golden Bug Net properly evolves the spawned pokemon based on level
- **[FIXED]** Old Gateau no longer saves its pregenArgs improperly
- **[FIXED]** Bug-Type Superfan can give RARE_FORM_CHANGE_ITEMs now (Mega stones, Max Mushrooms)
- **[FIXED]** Fun and Games: Wobbuffet can no longer be Roar/Dragon Tailed
- **[FIXED]** Old saves no longer load the Mysterious Challengers ME in error when they should not
- **[FIXED]** Uncommon Breed should no longer crash the game if it can't find species egg moves for the spawned pokemon

## Not Fixed / Low Priority

- There are still some extremely minor edge cases where egg gacha exploit is still abusable
- Visual bug with EXP bar on Weird Dream refusal option
- Player can still switch pokemon via U-Turn / Self-Destruct in Fun and Games (don't think this should be fixed)

### CODE REVIEW

File: src/data/mystery-encounters/encounters/absolute-avarice-encounter.ts [hunk 1]
```diff
@@ -262,15 +263,13 @@ export const AbsoluteAvariceEncounter: MysteryEncounter =
 
           // Provides 1x Reviver Seed to each party member at end of battle
           const revSeed = generateModifierType(scene, modifierTypes.REVIVER_SEED);
+          encounter.setDialogueToken("foodReward", revSeed?.name ?? "Reviver Seed");
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/absolute-avarice-encounter.ts:266]: Uhm. this seems like hard-coded english text. I know it's a fallback but...  it's basically bypassing any changes ever to be made to the i18n key of `reviverSeed`

File: src/data/mystery-encounters/encounters/bug-type-superfan-encounter.ts [hunk 2]
```diff
@@ -177,9 +178,11 @@ const MISC_TUTOR_MOVES = [
   Moves.U_TURN
 ];
 
+const WAVE_LEVEL_BREAKPOINTS = [30, 50, 70, 100, 120, 140, 160];
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/bug-type-superfan-encounter.ts:181]: Could we have an brief explanation what those are?
e.g.
```ts
/** Defines all the waves it changes something something bla.. foo.. */
const WAVE_LEVEL_BREAKPOINT = [...]
```

File: src/data/mystery-encounters/encounters/bug-type-superfan-encounter.ts [hunk 3]
```diff
@@ -216,11 +219,45 @@ export const BugTypeSuperfanEncounter: MysteryEncounter =
         female: true,
       });
 
+      let beedrillKeys, butterfreeKeys;
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/bug-type-superfan-encounter.ts:222]: NIT: Missing types

The expected type actually was never defined so you might have to add it.
Cause putting a 
```ts
let beedrillKeys: {spriteKey: string, something: string}
```
there seems stupid

File: src/data/mystery-encounters/encounters/bug-type-superfan-encounter.ts [hunk 4]
```diff
@@ -216,11 +219,45 @@ export const BugTypeSuperfanEncounter: MysteryEncounter =
...
+      if (scene.currentBattle.waveIndex < WAVE_LEVEL_BREAKPOINTS[3]) {
+        beedrillKeys = getSpriteKeysFromSpecies(Species.BEEDRILL, false);
+        butterfreeKeys = getSpriteKeysFromSpecies(Species.BUTTERFREE, false);
+      } else {
+        beedrillKeys = getSpriteKeysFromSpecies(Species.BEEDRILL, false, 1);
+        butterfreeKeys = getSpriteKeysFromSpecies(Species.BUTTERFREE, false, 1);
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/bug-type-superfan-encounter.ts:228]: NIT (super duper): It wouldn be nice to give the `formInde: 1` an actual name by using a constant
Something like `MEGA_BEEDRILL_FORM_INDEX = 1`

File: src/data/mystery-encounters/encounters/clowning-around-encounter.ts [hunk 5]
```diff
@@ -336,8 +336,8 @@ export const ClowningAroundEncounter: MysteryEncounter =
               .filter(move => move && !originalTypes.includes(move.getMove().type) && move.getMove().category !== MoveCategory.STATUS)
               .map(move => move!.getMove().type);
             if (priorityTypes?.length > 0) {
-              priorityTypes = [...new Set(priorityTypes)];
-              randSeedShuffle(priorityTypes);
+              priorityTypes = [...new Set(priorityTypes)].sort();
+              priorityTypes = randSeedShuffle(priorityTypes);
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/clowning-around-encounter.ts:340]: Q: What's the benefit of the `sort()` call before shuffling?

File: src/data/mystery-encounters/encounters/clowning-around-encounter.ts [hunk 6]
```diff
@@ -494,6 +494,10 @@ function generateItemsOfTier(scene: BattleScene, pokemon: PlayerPokemon, numItem
   }
 
   for (let i = 0; i < numItems; i++) {
+    if (pool.length === 0) {
+      // Stop generating new items if somehow runs out of items to spawn
+      return;
+    }
     const randIndex = randSeedInt(pool.length);
     const newItemType = pool[randIndex];
     let newMod;
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/clowning-around-encounter.ts:503]: NIT: Missing type

File: src/locales/en/mystery-encounters/safari-zone-dialogue.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
   "intro": "It's a safari zone!",
   "title": "The Safari Zone",
-  "description": "There are all kinds of rare and special Pokémon that can be found here!\nIf you choose to enter, you'll have a time limit of @[TOOLTIP_TITLE]{{{numEncounters}} wild encounters} where you can try to catch these special Pokémon.\n\nBeware, though. These Pokémon may flee before you're able to catch them!",
+  "description": "There are all kinds of rare and special Pokémon that can be found here!\nIf you choose to enter, you'll have a time limit of@[TOOLTIP_TITLE]{ {{numEncounters}} wild encounters} where you can try to catch these special Pokémon.\n\nBeware, though. These Pokémon may flee before you're able to catch them!",
```
Comment (flx-sta) [inline src/locales/en/mystery-encounters/safari-zone-dialogue.json:4]: Is this something we have to bring up with other languages too?

File: src/modifier/modifier.ts [hunk 8]
```diff
@@ -12,16 +13,15 @@ import * as Utils from "../utils";
 import { getBerryEffectFunc, getBerryPredicate } from "../data/berry";
 import { BattlerTagType } from "#enums/battler-tag-type";
 import { BerryType } from "#enums/berry-type";
-import { StatusEffect, getStatusEffectHealText } from "../data/status-effect";
+import { getStatusEffectHealText, StatusEffect } from "../data/status-effect";
```
Comment (flx-sta) [inline src/modifier/modifier.ts:16]: NIT: use `#app/...` import too

File: src/data/mystery-encounters/encounters/clowning-around-encounter.ts (hunk 5 above)
Comment (PigeonBar) [reply src/data/mystery-encounters/encounters/clowning-around-encounter.ts:340]: https://github.com/pagefaultgames/pokerogue/issues/4372#issuecomment-2367182706

It prevents the player from manipulating the randomly generated type by swapping the order of a Pokemon's moveset.

Comment (ben-lear) [reply src/data/mystery-encounters/encounters/clowning-around-encounter.ts:340]: image

File: src/locales/en/mystery-encounters/safari-zone-dialogue.json (hunk 7 above)
Comment (ben-lear) [reply src/locales/en/mystery-encounters/safari-zone-dialogue.json:4]: I checked other languages, they have not been updated to even have the TOOLTIP_TITLE styling stuff yet on that text.

File: src/data/mystery-encounters/encounters/bug-type-superfan-encounter.ts (hunk 4 above)
Comment (ben-lear) [reply src/data/mystery-encounters/encounters/bug-type-superfan-encounter.ts:228]: gonna add a comment but not change to a const

**Comment by PigeonBar** [comment]
I think the "passive Suction Cups" fix for phazing Wobbuffet isn't working because the passive doesn't activate due to Wobbuffet not being a Boss:

https://github.com/user-attachments/assets/d3b2c074-f0e4-4bc8-90f3-1ee2d008fcf0

(Besides, the fix wouldn't work if the player uses something like Gastro Acid.)

**Comment by ben-lear** [comment]
chrome_9ZcNOiScDX

Real fix to Fun and Games force switch

**Comment by flx-sta** [comment]
We should probably (in the future?) add a `Battle.isMysteryEncounter()` helper method.
image

File: src/data/mystery-encounters/encounters/training-session-encounter.ts [hunk 9]
```diff
@@ -322,27 +319,23 @@ export const TrainingSessionEncounter: MysteryEncounter =
...
+              if (!isNullOrUndefined(rootFusionSpecies)
+                && speciesStarters.hasOwnProperty(rootFusionSpecies)
+                && !!scene.gameData.dexData[rootFusionSpecies].caughtAttr) {
+                scene.gameData.starterData[rootFusionSpecies].abilityAttr |= playerPokemon.fusionAbilityIndex !== 1 || playerPokemon.fusionSpecies?.ability2
+                  ? 1 << playerPokemon.fusionAbilityIndex
+                  : AbilityAttr.ABILITY_HIDDEN;
               }
             } else {
               playerPokemon.abilityIndex = abilityIndex;
-              if (speciesStarters.hasOwnProperty(playerPokemon.species.speciesId)) {
```
Comment (MokaStitcher) [inline src/data/mystery-encounters/encounters/training-session-encounter.ts:336]: for fusions we unlock the ability (if applicable), but we don't unlock it for non fusions anymore? I'm a bit confused

Comment (ben-lear) [reply src/data/mystery-encounters/encounters/training-session-encounter.ts:336]: for non-fusions, the ability is unlocked as part of `setPokemonCaught()`, down below

**Review by flx-sta — DISMISSED** [review]
code: lgtm 🚀 
play: verified 2 encounters for fixes 
- Fun & Games  ✅ 
- Weird dream  ✅

File: src/data/mystery-encounters/utils/encounter-pokemon-utils.ts [hunk 10]
```diff
@@ -757,17 +757,19 @@ const GOLDEN_BUG_NET_SPECIES_POOL: [Species, number][] = [
...
   const totalWeight = GOLDEN_BUG_NET_SPECIES_POOL.reduce((a, b) => a + b[1], 0);
   const roll = randSeedInt(totalWeight);
 
   let w = 0;
   for (const speciesWeightPair of GOLDEN_BUG_NET_SPECIES_POOL) {
     w += speciesWeightPair[1];
     if (roll < w) {
-      return getPokemonSpecies(speciesWeightPair[0]);
+      const initialSpecies = getPokemonSpecies(speciesWeightPair[0]);
+      return getPokemonSpecies(initialSpecies.getSpeciesForLevel(level, true));
```
Comment (DayKev) [inline src/data/mystery-encounters/utils/encounter-pokemon-utils.ts:772]: What is the circumstance these Pokémon are being generated for? Are they wild, enemy trainer, [...]?

Comment (ben-lear) [reply src/data/mystery-encounters/utils/encounter-pokemon-utils.ts:772]: wild

Comment (ben-lear) [reply src/data/mystery-encounters/utils/encounter-pokemon-utils.ts:772]: https://github.com/pagefaultgames/pokerogue/blob/34bf0069caeb625d226fb75ae29ce5151908ef84/src/phases/encounter-phase.ts#L108

**Review by DayKev — DISMISSED** [review]

**Comment by PigeonBar** [comment]
I think the fix for not updating dex entries for rental Pokemon is causing dex entry updates to fail for ALL caught Pokemon. In this example, I catch Raikou but fail to unlock it as a starter:

https://github.com/user-attachments/assets/d58057e0-21c1-4919-adcf-20b35f873005

**Comment by ben-lear** [comment]
Catches (and hatches) update dex properly, as normal again.

chrome_gN3sJYqlOJ

Form Changes etc. for "rental" pokemon (Deoxys from Weird Dream) do not unlock the mon in dex (accidentally stopped recording so 2 videos)

chrome_BGERwntIaD

chrome_pxivUxDDpc

**Review by DayKev — APPROVED** [review]

**Review by flx-sta — APPROVED** [review]

**Pull request merged**